receiver/awscloudwatchreceiver/logs.go

```diff
@@ -0,0 +1,353 @@
+	request(limit int, nextToken string, st, et *time.Time) *cloudwatchlogs.FilterLogEventsInput
+}
+
+func newLogsReceiver(cfg *Config, logger *zap.Logger, consumer consumer.Logs) *logsReceiver {
+	namedPolls := []groupNames{}
+	prefixedPolls := []groupPrefix{}
+	for logGroupName, sc := range cfg.Logs.Groups.NamedConfigs {
+		for _, prefix := range sc.Prefixes {
+			prefixedPolls = append(prefixedPolls, groupPrefix{
+				logGroup: logGroupName,
+				prefix:   prefix,
+			})
+		}
+
+		namedPolls = append(namedPolls, groupNames{
+			logGroup:    logGroupName,
+			streamNames: sc.Names,
+		})
+	}
+
+	// safeguard from using both
+	autodiscover := cfg.Logs.Groups.AutodiscoverConfig
+	if len(cfg.Logs.Groups.NamedConfigs) > 0 {
+		autodiscover = nil
+	}
+
+	return &logsReceiver{
+		region:              cfg.Region,
+		profile:             cfg.Profile,
+		consumer:            consumer,
+		maxEventsPerRequest: cfg.Logs.MaxEventsPerRequest,
+		imdsEndpoint:        cfg.IMDSEndpoint,
+		autodiscover:        autodiscover,
+		pollInterval:        cfg.Logs.PollInterval,
+		namedPolls:          namedPolls,
+		prefixedPolls:       prefixedPolls,
+		logger:              logger,
+		wg:                  &sync.WaitGroup{},
+		doneChan:            make(chan bool),
+	}
+}
+
+func (l *logsReceiver) Start(ctx context.Context, host component.Host) error {
+	l.logger.Debug("starting to poll for Cloudwatch logs")
+
+	if l.autodiscover != nil {
+		namedPolls, prefixedPolls, err := l.discoverGroups(ctx, l.autodiscover)
+		if err != nil {
+			return fmt.Errorf("unable to auto discover log groups using prefix  %s: %w", l.autodiscover.Prefix, err)
+		}
+		l.namedPolls = namedPolls
+		l.prefixedPolls = prefixedPolls
+	}
+	l.wg.Add(1)
+	go l.startPolling(ctx)
+	return nil
+}
+
+func (l *logsReceiver) Shutdown(ctx context.Context) error {
+	l.logger.Debug("shutting down logs receiver")
+	close(l.doneChan)
+	l.wg.Wait()
+	return nil
+}
+
+func (l *logsReceiver) startPolling(ctx context.Context) {
+	defer l.wg.Done()
+
+	t := time.NewTicker(l.pollInterval)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-l.doneChan:
+			return
+		case <-t.C:
+			err := l.poll(ctx)
+			if err != nil {
+				l.logger.Error("there was an error during the poll", zap.Error(err))
+			}
+		}
+	}
+}
+
+func (l *logsReceiver) poll(ctx context.Context) error {
+	var errs error
+	for i := range l.namedPolls {
+		np := l.namedPolls[i]
+		if err := l.pollForLogs(ctx, np.logGroup, &np); err != nil {
+			errs = multierr.Append(errs, err)
+		}
+	}
+	for i := range l.prefixedPolls {
+		pp := l.prefixedPolls[i]
+		if err := l.pollForLogs(ctx, pp.logGroup, &pp); err != nil {
+			errs = multierr.Append(errs, err)
+		}
+	}
+	return errs
+}
+
+func (l *logsReceiver) pollForLogs(ctx context.Context, logGroup string, pc pollConfig) error {
+	err := l.ensureSession()
+	if err != nil {
+		return err
+	}
+	nextToken := aws.String("")
+	startTime := time.Now().Add(-l.pollInterval)
+	endTime := time.Now()
```

Comment:
You need to keep track of the exact time at which the previous request was made. Otherwise, you are losing and/or duplicating logs based on timing issues.